### ISSUE TITLE
PR: Mispelling inferences

### ISSUE THREAD

**Issue by amilajack** [eslint/eslint#6794]
Would a PR for something like this be welcome:

```js
// Simple example of making inferences through spelling
const some = Promis // <- misspelled Promise
```

When linting this code, eslint would output something like so:

```
Unexpected token 'Promis'. Did you mean 'Promise'?
```

instead of:
```
Parsing Error: Unexpected token.
```

**Comment by ZakariaRidouh** [comment]
This would be very useful to catch small issues 😄

**Comment by platinumazure** [comment]
I'm not sure I understand why the example would give a parse error. Are you using ecmaVersion: 6 or later?

The no-undef rule should already flag this sort of thing.

**Comment by nzakas** [comment]
@amilajack there's really not enough information here for us to evaluate. Your example doesn't work that way so we can't be sure what you are proposing. Are you proposing a change to a single rule? All rules? Something else?

**Comment by amilajack** [comment]
I'm proposing to add functionality that can 'guess' the correct spelling of variables/code.

**Comment by kaicataldo** [comment]
@amilajack Are you envisioning this as a linting rule? Something that happens at the parser level? What if you intended to use `Promis` and this rule wouldn't let you? Interested to see how you envision this could work.

**Comment by amilajack** [comment]
@kaicataldo my current linter configuration (which extends the airbnb config) throws a parser error when a variable is not defined. It believe it ignores globally defined variables. That way, eslint knows that this is indeed a spelling mistake (because i am not importing `Promis`).

**Comment by nzakas** [comment]
@amilajack I'm sorry, but you haven't answered any of the questions that the team has asked. We really can't evaluate this without very specific information. Please read over our comments again and answer the questions that were asked.

**Comment by amilajack** [comment]
> Are you proposing a change to a single rule? All rules? Something else?

Ideally, this can be configured by a rule. I'm sure that some work needs to be done with the parser in order to manually handle Parser errors.

> Are you proposing a change to a single rule? All rules? Something else?

Ideally, there would be a single rule, say something like:
```
"did-you-mean": [2, { "ignore": ["VariableToIgnore"] }]
```

> Are you envisioning this as a linting rule?
Yes

> Something that happens at the parser level? 
If rules do not have access to handle Parser errors then there will need to be some changes to the parser.

This is an idea from the ruby community. [This ruby package](https://github.com/yuki24/did_you_mean) is a perfect example of this idea.

**Comment by ilyavolodin** [comment]
@amilajack Could you provide a complete example to reproduce this? As well as your configuration file? Undefined variables do not throw "Parser error". Parser errors is for invalid syntax that just can't be parsed into AST. Undefined variables, while a cause of runtime issues, are perfectly legal syntax. `no-undef` rule should catch those and display appropriate message (without the guessing part).

**Comment by amilajack** [comment]
@ilyavolodin Sorry that was a mistake. Its actually the`no-undef` rule that does this. So there wouldnt need to be any work done with the parser.

**Comment by ilyavolodin** [comment]
In that case, `no-undef` already notifies you about usage of undefined variables. What's the advantage of us guessing what you might've meant if the rule already points you to the location of the wrong variable? Chances are, we can't really guess correctly what was supposed to be there in the first place. User would know much better what he was planning to type there. I don't see this is useful enough to justify complexity that this would add to one of the most widely used rules out there. Overall I'm :-1:  on this, sorry.

**Comment by alberto** [comment]
Closing since there seems to be consensus against this

**Comment by platinumazure** [comment]
@amilajack Just want to note that even though we don't feel we can reasonably add this to `no-undef` in core, ESLint is completely pluggable and you have the freedom to create your own rule.

In fact, you could create a plugin that pulls in something like sindresorhus/globals as a dependency, and use that to seed a suggested words list (while allowing users to write in their own words). Then you could write rules which look at identifiers and/or string literals, tokenize them from camelcase or underscore-delimited, and check each word against your word list. I think that would actually make an awesome plugin! But it doesn't necessarily fit in as a core rule.

If you want to start that project and you need help developing the plugin, you can always stop by our [Gitter chat](https://gitter.im/eslint/eslint).
